flax/core/frozen_dict.py

```diff
@@ -244,7 +248,8 @@ def copy(
     return x.copy(add_or_replace)
   elif isinstance(x, dict):
     new_dict = jax.tree_map(lambda x: x, x)  # make a deep copy of dict x
-    new_dict.update(add_or_replace)
+    if add_or_replace is not None:
+      new_dict.update(add_or_replace)
```

Comment:
```suggestion
    if isinstance(add_or_replace, dict):
      add_or_replace = jax.tree_map(lambda x: x, add_or_replace)
    new_dict.update(add_or_replace)
```

Reply:
I'm not sure I understand the logic here. 
Using your MappingProxy idea, I can just remove this change, and wouldn't need to touch those lines.